File: backend/app/schemas/agent.py
"""对话契约：MessageIn（入参校验）与 PlanStepOut / TurnOut（响应）。

TurnOut.status 是 RunStatus 五值但只表示规划本身成败——执行状态在
ToolRun（步骤卡片经 useRun 读取），两层抽象不混用。
"""
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models.agent_run import AgentRun
from app.models.tool_run import RunStatus
from app.tools import label_of

MAX_MESSAGE = 1000


class MessageIn(BaseModel):
    text: str = Field(min_length=1, max_length=MAX_MESSAGE)

    @field_validator("text")
    @classmethod
    def text_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("消息不能为空白")
        return value


class PlanStepOut(BaseModel):
    tool: str
    label: str  # 注册表文案；步骤卡与历史列表共用
    run_id: uuid.UUID | None = None  # 校验被拒的步骤没有 run


class TurnOut(BaseModel):
    id: uuid.UUID
    revision: int  # 规划时的会话修订号（规划基准）
    goal: str
    reply: str
    status: RunStatus
    error: str | None
    created_at: datetime
    steps: list[PlanStepOut]

    @classmethod
    def of(cls, record: AgentRun) -> "TurnOut":
        steps = [
            PlanStepOut(
                tool=str(step.get("tool", "")),
                label=label_of(str(step.get("tool", ""))),
                run_id=uuid.UUID(step["run_id"]) if step.get("run_id") else None,
            )
            for step in (record.plan or [])
        ]
        return cls(
            id=record.id,
            revision=record.revision,
            goal=record.goal,
            reply=record.reply,
            status=record.status,
            error=record.error,
            created_at=record.created_at,
            steps=steps,
        )

File: frontend/src/api/agent.ts
/** 对话 API：Turn / PlanStep 类型 + agentApi；全部走统一 client（Cookie 自动携带）。 */
import { api } from '@/api/client'
import type { RunStatus } from '@/api/runs'

/** 一步工具调用：run_id 可空——校验被拒的步骤没有 run。 */
export interface PlanStep {
  tool: string
  label: string
  run_id: string | null
}

/** 一轮对话：status 只表示规划本身成败；执行状态在 PlanStep.run_id 对应的 Run 里。 */
export interface Turn {
  id: string
  revision: number
  goal: string
  reply: string
  status: RunStatus
  error: string | null
  created_at: string
  steps: PlanStep[]
}

export const agentApi = {
  turns: (sessionId: string) => api.get<Turn[]>(`/sessions/${sessionId}/messages`),
  send: (sessionId: string, text: string) =>
    api.post<Turn>(`/sessions/${sessionId}/messages`, { text }),
}
